module/metrics.go

```diff
@@ -715,7 +715,7 @@ type HeroCacheMetrics interface {
 	// each key belongs to an existing (key, entity) pair.
 	// Hence, adding a new key to that bucket will replace the oldest valid key inside that bucket.
 	// Note: in context of HeroCache, the key corresponds to the identifier of its entity.
-	OnEntityEjectionDueToEmergency()
+	OnEntityEjectionDueToEmergency(ejectedEntity flow.Entity)
```

Comment:
Let's not use metric collectors to track down the details of the entities. The reason is it reduces the cohesion of the interface; i.e., rather than focusing on a single job, the metrics collection is now turned into a multi-task interface; it updates the metrics collection and records the details of certain entities (ejected ones in this case). This makes the maintainability and clarity of the code quite daunting over the long term. 

Alternatively, I suggest we use a dependency injection pattern and inject tracer objects into HeroCache. A tracer object is supposed to be called upon certain events, such as the entity ejection. 

A raw draft of refactoring with such tracers will look like:
```go
type Tracer interface {
	EntityEjectionDueToEmergency(ejectedEntity flow.Entity)
	EntityEjectionDueToFullCapacity(ejectedEntity flow.Entity)
}
```

Such a tracer can be set at the initialization time through an option (if it is not set, it remains `nil`): 
```go
func WithTracer(t Tracer) CacheOpt {
    return function(c *Cache) {
                c.tracer = t
    }
}
```

Then upon such events, we can check the existence of the tracer and call it if one has set; 
```go
ejectedEntity := c.invalidateEntity(b, slotToUse)
if c.tracer != nil {
   c.tracer.EntityEjectionDueToEmergency(ejectedEntity)
}
c.collector.OnEntityEjectionDueToEmergency()
```

Reply:
I was considering this approach but a few things stopped me:
1. While this interface is used only by metrics it seems like a general enough interface and there potentially could be different implementors, for instance logging or observing rate of failures.
2. The generality of interface motivated me to add extra parameter(`ejectedEntity`) since it's quite useful to have extra information about ejection. 

I could change it the way you propose if you insist, but the current interface seems general enough to support usage that I have implemented. Let me know your thoughts.

Reply:
Thanks for suggestions. https://github.com/onflow/flow-go/pull/3965/commits/58733d183ad57122ef146caed2dd9c6593496808 let me know if you want something changed. I have used your example to implement it.